### DISCUSSION TITLE
IDE ideas

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#33] [Ideas]
My one-line pitch for kons-9 is **"An IDE for 3D production."**

Which invites the question of what features such an IDE would entail.

kons-9 is a system for developing 3D content by using graphical interaction and programming means. In this sense it is also an IDE for developing Common Lisp code. The potential of the system lies in the interplay of these two approaches.

Possibilities for the architecture of the system range from free-form interaction using Emacs/Slime, where the user can do anything programmatically, to more structured approaches in terms of well-managed transactions that can be logged.

It is going to be fun exploring these ideas.

**Comment by omarbassam88** [comment]
Do you mean to add Code Editing Capabilities for Lisp code inside kons-9 without sacrificing using Emacs (or other editor) to interact with the IDE? Like how for Example Houdini does for Vex, you can edit Vex inside of Houdini but you can also use VS Code to edit your Vex code.

**Comment by kaveh808** [comment]
My goal is to do much more than what Vex lets you do, which if I understand properly, is to write scripts inside nodes.

A major advantage of our system being implemented in Common Lisp is that our implementation language is our scripting language. Kind of as if Houdini easily let you do anything you wanted in C++ rather than having (multiple) separate scripting languages.

So we want ways of making writing the code inside our system as easy as possible, hence an IDE, or IDE-like features.

**Comment by awolven** [comment]
I don't want to beat a dead horse here, but I have ported the CMUCL
emacs-compatible text editor "Hemlock" to the 3d graphics world and it
works.  The future is here now.  You open lisp code in Hemlock, hit Ctrl-C
Ctrl-C and your 3d graphics object updates shape in the graphics window.
You can also use GNU Emacs with sly/sime while you're running Hemlock in
the graphics window environment.  It currently has some dependencies on
windows, which could be easily overcome, just for the features of "dired"
and "shell", and has minimal dependencies on event handling (which is
glfw) and text and quads which could be provided by the full-lisp ui Alloy,
and has no dependencies at all on whether you use OpenGL or not.  It took
me about six months to port.  It also could be easily extended to provide
for text-widgets.

**Reply by mikelevins** [reply]
Hemlock is a solid option. Both the text editor in CCL's Cocoa IDE and the text editor in the Lispworks graphical UI are forks of Hemlock.

**Comment by awolven** [comment]
My fork of Hemlock is thread safe and only uses operating system calls for
file system or pipe stuff.  I doubt you would be able to get an incremental
redisplay hemlock working with OpenGL, both of which Lispworks (commercial)
and CCL are.

**Comment by kaveh808** [comment]
Nice. It would be very cool if we could use Hemlock as a text widget in our UI. I would like to have Emacs-like text editing capabilities in our UI.

**Reply by mikelevins** [reply]
Hemlock uses string-based commands that remind me more of Zmacs than of GNU Emacs.

**Comment by awolven** [comment]
I would be happy to do the legwork for that, I just need you to decide what
do you want to use for the font and simple windowing features, and which os
because it will have to have the ability to do file system tasks.
Currently it caters to MS Windows, though that was ported from unix albeit
the libraries it used for unix are now defunct.

**Comment by kaveh808** [comment]
Our goals are to support MacOS, Linux, and Windows. I like Monaco as my Emacs font.

Not sure about windowing features yet. We have not yet decided on a GUI architecture.

**Comment by awolven** [comment]
There are various systems, most which depend on freetype to render the
fonts into textures, where the glyphs are mapped onto quads.  One of which
would be necessary to complete the project.  I think both |3b| and Shinmera
have these features in their systems, and I am using ImGui for that.

**Comment by JMC-design** [comment]
Can you provide access to your fork yet?  I'd like to see how this looks/works.
3b is doing (m)sdf generation of font atlases, and Shinmera is just using those font atlases.
Freetype isn't in this equation because that's a font rasterizer while sdf's are generated from paths.
Unless you are using rasterized bitmaps?  I'd still like to see that handle by CL, kinda why I wrote https://github.com/JMC-design/font . Still need to clean up my Type 1 backend, and finish the bdf/pcf ones.  When Scymtym finally finishes Type 3 we'll have even more options.  
I don't think it's a good idea to depend on system libraries for the font rendering because then your apps look different on every OS, with the worst being windows(god that zoom is horrible). But others might not feel that strongly (look at all the windows users!).

**Reply by mikelevins** [reply]
You kind of have to pick a design philosophy. You can either use system APIs and resources so that your stuff looks native to the platform, or you can use your own portable stuff so that everything looks native to your app, and maybe out of place on one platform or another.

Whichever way you go, you will make someone unhappy. Windows-only users will be unhappy if things don't look like Windows (they love to hate on macOS' "fuzzy" fonts, for example). macOS-only users will be unhappy if things don't look like Cocoa and Quartz. Linux on the desktop is sort of a free-for all, so everybody sort of  expects every app to look different, more or less.

If you think users will care more about the app being consistent than matching the platform, then use your own stuff. If you care about placating platform nativists, use platform-native stuff.

**Reply by JMC-design** [reply]
sure, but I thought part of the goal of this was growing the CL ecosystem?  Or maybe I just placed that in there because of my own goals, my bad if I did.

**Reply by mikelevins** [reply]
No, Kaveh definitely wants to do that, from other comments he made. I'm not saying anything against that goal. I'm saying there's no way to make everyone happy, so you just have to decide which set of people you think it's better to annoy. 

(Or you can ignore those considerations, understanding that annoying someone is inevitable, and just pick what you'd rather work with.)

**Reply by kaveh808** [reply]
Yes, I definitely want to do that. I want a CL all the way down (as much as practical) portable user interface system that looks the same on all platforms. It will be a great benefit to the CL ecosystem as well.

**Reply by mikelevins** [reply]
"looks the same on all platforms" sounds like a decision about the design philosophy. It implies what JMC suggested: avoid platform-specific libraries and assets.

It means that the set of complaints you've decided to ignore is the ones from the single-platform nativists. (No judgment from me; you have to pick either the platform nativists or the app nativists; I'm good either way you want to go.)

**Reply by JMC-design** [reply]
though tbf, editors are usually monospace bitmap fonts. So no big difference there.
It would be nice if some vector graphics wizard could implement more performant vector rasterization in Lisp.
Looking at improving cl-vectors is on my todo list for the far future, but I'm no wizard.

**Comment by awolven** [comment]
I'm open to using a different font rasterizer than freetype, I do prefer a
CL solution though.  What would be really cool while we're on the subject
is if someone came up with a vector font library, where you could
dynamically scale the font without losing the crispness.  As far as look
and feel I vote that, as a CL community project, we create our own look and
feel which is the same on all platforms.

For any of the folks that want a preview of hemlock, send your public key
to awolven at gmail.com and I will send you the access link.

**Reply by JMC-design** [reply]
If you mean on opengl, generation of (m)sdf texture atlases is exactly what 3b is working on.  That comes free if you use Alloy I think, since it's what Shinmera uses. 
If you mean just being able to rasterize vector paths at whatever size you want, we've already got that as well. If you try the font-zpb-ttf backend you can get a cl-vector path from a TTF by (glyph:path (font:glyph character some-open-ttf-font)) scale as you wish and render, will be the same for Type 1 when clean up the parser to use the same font structure as realized fonts(what I call vector fonts prescaled to a certain size)..

lol, as for my public key, I believe I left it in 90s.

**Comment by starkaiser** [comment]
As an idea, maybe we can use Lem https://github.com/lem-project/lem and somehow embed it in the program. It is an IDE written in Common Lisp that supports multithreading. 
It would be cool if kons-9 was a "virtual lisp machine" and you could make snapshots. For example, let's say you study generative design. You can make a snapshot which implements a genetic algorithm and another that implements other type of algorithm, run the two snapshots separately and compare the results. This would be handy when you want to explore different ideas.Also, if you made changes and break the system, you could go back to a previous snapshot that works. For now Emacs/SLIME is fine, but in the future I think we should implement the editor in the kons-9 system. It is easier to have everything in a place, especially for non-developers
 and newcomers.
I know that Pharo is a combination of IDE and virtual machine and you can have multiple Pharo snapshots and launch the one you need. Maybe we can make something along those lines. Just an idea

**Reply by JMC-design** [reply]
Lem is console based.  Nobody has written a terminal abstraction yet. It's far down on my todo list.

**Reply by mikelevins** [reply]
Agreed.

@awolven's hemlock port is a better bet.

**Reply by JMC-design** [reply]
Hard to tell without actually seeing it.

**Reply by mikelevins** [reply]
I mean, obviously. Working on it. I’ll let you know.

**Reply by kaveh808** [reply]
Love the idea.

This completely lines up with my vision for the system. I want it to be a full-fledged IDE with graphics, as mentioned in the Readme.

If we go as far as a virtual Lisp Machine, that would be fabulous.

**Reply by mikelevins** [reply]
I want that, too. So much so that I already have a different work in progress aimed at that goal. If kons-9 develops well enough, I'll frankly abandon that effort and concentrate on kons-9 instead.

It's a high hill to climb, though. 

To take one example of why, I want to revive my circa-2000 programmable word processor, Alpaca. I tried to work out a license with Lispworks that would allow it, but we couldn't get past their "nothing that can be used as a Lisp development environment" restriction.

The original was written in CCL with the Objective-C bridge, and so was Mac-only. I'd rather have it cross platform. It needs to address general-purpose text rendering, though--not just for a 3D UI, but for proper text editing and word processing.

Here's a nice discussion of what that entails:

https://faultlore.com/blah/text-hates-you/

**Reply by aykaramba** [reply]
The whole set of conversations around this project have been mind blowing.  A 3d accelerated Hemlock is one hell of an idea.  A virtual machine takes it to a whole other level.  All of the discussions are simply great reading.

As a side point, I assume everyone knows there is a maintained common lisp OS that Froggey from #lisp on irc.libera.chat just dropped out of the blue on everyone a few years ago:  https://github.com/froggey/Mezzano

Mezzano runs as a virtual machine and apparently can be installed on some type of bare metal or hopes to be some day, but never looked into it that far.

Not sure if in any way applicable but a good looksee for those that have not seen it before.  It's quite fun to play with in a vm.

**Reply by JMC-design** [reply]
huh, looks like lem does support different frontends.

**Reply by mikelevins** [reply]
Send a link; I'm interested (or if you don't feel like it, don't worry; I'll search it out myself).

Edit: found this one: https://github.com/gregcman/lem-opengl

**Reply by JMC-design** [reply]
That's interesting. Uses yet another modern GL abstraction.
